litert/vendors/qualcomm/core/transformation/mha_to_sha.cc

```diff
@@ -786,12 +777,296 @@ size_t OptimizeMHAFastVlmPrefill(
               ops.begin() + start_index + pattern_size);
     return step_size;
   }
+  QNN_LOG_WARNING(
+      "[G2G] Validation failed. Rolling back to the original graph.");
+  return 1;
+}
+namespace {
+
+bool OptimizeMHATinyGemmaPrefill(
+    std::vector<OpWrapper>& new_ops, TensorPool& tensor_pool,
+    const OpWrapper& mul, const OpWrapper& transpoe_0,
+    const OpWrapper& reshape_0, const OpWrapper& matmul_k0,
+    const OpWrapper& matmul_k1, const OpWrapper& concat, const OpWrapper& add_0,
+    const OpWrapper& softmax, const OpWrapper& slice_0,
+    const OpWrapper& slice_1, const OpWrapper& matmul_v0,
+    const OpWrapper& matmul_v1, const OpWrapper& add_1,
+    const OpWrapper& reshape_1, const OpWrapper& transpose_1,
+    const OpWrapper& reshape_2) {
+  const auto is_connected =
+      [](const OpWrapper& output, size_t output_tensor_index,
+         const OpWrapper& input, size_t input_tensor_index) -> bool {
+    return output.GetOutputTensor(output_tensor_index) ==
+           input.GetInputTensor(input_tensor_index);
+  };
+  if (!(is_connected(mul, 0, transpoe_0, 0) &&
+        is_connected(transpoe_0, 0, reshape_0, 0) &&
+        is_connected(reshape_0, 0, matmul_k0, 0) &&
+        is_connected(reshape_0, 0, matmul_k1, 0) &&
+        is_connected(matmul_k0, 0, concat, 0) &&
+        is_connected(matmul_k1, 0, concat, 1) &&
+        is_connected(concat, 0, add_0, 0) &&
+        is_connected(add_0, 0, softmax, 0) &&
+        is_connected(softmax, 0, slice_0, 0) &&
+        is_connected(softmax, 0, slice_1, 0) &&
+        is_connected(slice_0, 0, matmul_v0, 0) &&
+        is_connected(slice_1, 0, matmul_v1, 0) &&
+        is_connected(matmul_v0, 0, add_1, 0) &&
+        is_connected(matmul_v1, 0, add_1, 1) &&
+        is_connected(add_1, 0, reshape_1, 0) &&
+        is_connected(reshape_1, 0, transpose_1, 0) &&
+        is_connected(transpose_1, 0, reshape_2, 0))) {
+    return false;
+  }
+
+  QNN_LOG_INFO("[G2G] MHA optimization (TinyGemma Prefill)");
+  const auto& pattern_input = mul.GetInputTensor(0);
+  const auto& pattern_output = reshape_2.GetOutputTensor(0);
+
+  // Transpose
+  auto transpose_output_dims = transpoe_0.GetOutputTensor(0).GetDims();
+  auto& transpose_output =
+      tensor_pool.CloneNativeTensorFrom(pattern_input, transpose_output_dims);
+  auto& new_transpose_0 = EmplaceOpWithIO(
+      new_ops, transpoe_0, {const_cast<::qnn::TensorWrapper&>(pattern_input)},
```

Comment:
Can we remove this since we are already in `qnn` namespace?
```suggestion
      new_ops, transpoe_0, {const_cast<TensorWrapper&>(pattern_input)},
```

Reply:
Sure.